tools/ktx/ktx_main.cpp

```diff
@@ -0,0 +1,137 @@
+// Copyright 2022-2023 The Khronos Group Inc.
+// Copyright 2022-2023 RasterGrid Kft.
+// SPDX-License-Identifier: Apache-2.0
+
+
+#include "command.h"
+#include "stdafx.h"
+#include "utility.h"
+
+#include <iostream>
+#include <unordered_map>
+#include <string>
+#include <string_view>
+
+// -------------------------------------------------------------------------------------------------
+
+namespace ktx {
+
+/** @page ktxtools ktx
+@~English
+
+Unified CLI frontend for the KTX-Software library.
+
+@section ktxtools_synopsis SYNOPSIS
+    ktx @e command @e command_args...<br>
+    ktx [options]
+
+@section ktxtools_description DESCRIPTION
+    Unified CLI frontend for the KTX-Software library with sub-commands for specific operations.
+
+    The following commands are available:
+    <dl>
+        <dt>validate</dt>
+        <dd>
+            Prints information about KTX2 file.
+        </dd>
+        <dt>info</dt>
+        <dd>
+            Validates a KTX2 file.
+        </dd>
+        <dt>transcode</dt>
+        <dd>
+            @warning TODO KTX Tools P5: This section is incomplete
+        </dd>
+        <dt>encode</dt>
+        <dd>
+            @warning TODO KTX Tools P5: This section is incomplete
+        </dd>
+        <dt>extract</dt>
+        <dd>
+            @warning TODO KTX Tools P5: This section is incomplete
+        </dd>
+        <dt>create</dt>
+        <dd>
+            @warning TODO KTX Tools P5: This section is incomplete
+        </dd>
+        <dt>help</dt>
+        <dd>
+            @warning TODO KTX Tools P5: This section is incomplete
+        </dd>
+    </dl>
+
+    The following options are also available without a command:
+    @snippet{doc} ktx/command.h command options
+
+@section ktxtools_exitstatus EXIT STATUS
+    @b ktx @b info exits
+        0 on success,
+        1 on command line errors and
+        2 if the input file parsing failed.
+
+@section ktxtools_history HISTORY
+
+@par Version 4.0
+ - Initial version
+
+@section ktxtools_author AUTHOR
+    - Mátyás Császár [Vader], RasterGrid www.rastergrid.com
+    - Daniel Rákos, RasterGrid www.rastergrid.com
+*/
+
+class Tools : public Command {
+public:
+    Tools() {}
+    virtual ~Tools() {};
+
+public:
+    virtual bool processOption(argparser& parser, int opt) override {
+        (void) parser;
+        (void) opt;
+        // TODO KTX Tools P5: Parse --version
+        // TODO KTX Tools P5: Parse --help
+        return true;
+    }
+    virtual int main(int argc, _TCHAR* argv[]) override;
+};
+
+int Tools::main(int argc, _TCHAR* argv[]) {
+    (void) argc;
+    (void) argv;
+    // TODO KTX Tools P5: Print usage, Failure: incorrect sub command {commandName}
+    std::cerr << "Print usage, Failure: incorrect sub command {commandName}" << std::endl;
```

Comment:
For the program name, what do you expect to be output?

- "ktx" only
- the subcommand only
- "ktx <subcommand>"
- "ktx-<subcommand>"

Also, do we need to include the entire program path for "ktx"?

Reply:
My opinion would be to use the usual argv[0] as executable name and include any nested relevant sub-command even tho argv[0] is not always portable.

Reply:
I'd like to hear what @MarkCallow and @lexaknyazev would like to see in <program>.

Reply:
Just to be clear about why I'm asking is because there were plans to support in the future executable or shared library based extension commands, and this has an impact on that. In our discussion git was used as an example of such an extensible CLI tool which uses the `git-<command>` pattern for e.g. in the man pages, although it does not usually print the command as part of the error messages. So we need to know what should be the convention we follow here.

Reply:
Use argv[0] as the executable name and include the sub command name, unless of course the error occurs before a sub command is called. Extension commands via `git-<command>` do not inhibit the command name being included in error messages.

What do you mean "argv[0] is not always portable"?

Reply:
Nothing serious, just that its content is not guaranteed and its dependent on the process spawning context. So for example for deterministic tests it will have to wildcard/fuzzy match it

Reply:
Yes, argv[0] may even be NULL in certain special cases, but it's nothing serious.

So argv[0], even in subcommands, i.e. "ktx" independent of the subcommand being used?

Reply:
> So argv[0], even in subcommands, i.e. "ktx" independent of the subcommand being used?

Yes, e.g. "ktx validate".

Reply:
> Yes, e.g. "ktx validate".

So "argv[0]" ("ktx") for the main command "argv[0] argv[1]" (e.g. "ktx validate") for subcommands, i.e. replace argv[0] for downstream subcommands with "argv[0] argv[1]", as I understand. Thanks!

Reply:
I mark this as resolved for now and we will implement these in an upcoming PR (with the implementation of usages)